src/mqt/problemsolver/partialcompiler/qaoa.py

```diff
@@ -0,0 +1,145 @@
+
+            # Iterate over all possible two-qubit gates
+            for i in range(self.num_qubits):
+                # considered_following_qubits describes the number of to be considered following qubits
+                for j in range(i + 1, i + 1 + considered_following_qubits):
+                    if j >= self.num_qubits:
+                        break
+                    p = Parameter(f"a_{parameter_counter}")
+                    qc.rzz(p, i, j)
+                    # Sample whether the gate should be removed for the first layer
+                    if k == 0:
+                        if np.random.random() < self.sample_probability:
+                            remove_gates.append(p.name)
+                        else:
+                            remove_gates.append(False)
+                            qc_baseline.rzz(p, i, j)
+                    # For all other layers, check whether the gate should be removed
+                    elif remove_gates[parameter_counter - k * tmp_len]:
+                        remove_gates.append(p.name)
+                    else:
+                        remove_gates.append(False)
+                        qc_baseline.rzz(p, i, j)
+                    parameter_counter += 1
+
+            m = Parameter(f"b_{k}")
+
+            # Mixer Layer
+            qc.rx(2 * m, range(self.num_qubits))
+            qc_baseline.rx(2 * m, range(self.num_qubits))
+
+        qc.measure_all()
+        qc_baseline.measure_all()
+
+        return qc, qc_baseline, remove_gates
+
+    def compile_qc(self, baseline: bool = False, opt_level: int = 3) -> QuantumCircuit:
+        """Compiles the circuit"""
+        if baseline:
+            return transpile(self.qc_baseline, backend=self.backend, optimization_level=opt_level, seed_transpiler=42)
+        return transpile(self.qc, backend=self.backend, optimization_level=opt_level, seed_transpiler=42)
+
+    def get_to_be_removed_gate_indices(self) -> list[int]:
+        """Returns the indices of the gates to be removed"""
+        indices_parameterized_gates = []
+        for i, gate in enumerate(self.qc_compiled._data):
+            if (
+                gate.operation.name == "rz"
+                and isinstance(gate.operation.params[0], Parameter)
+                and gate.operation.params[0].name.startswith("a_")
+            ):
+                indices_parameterized_gates.append(i)
+        assert len(indices_parameterized_gates) == len(self.remove_gates)
+        indices_to_be_removed_parameterized_gates = []
+        for i in range(len(indices_parameterized_gates)):
+            if self.qc_compiled._data[indices_parameterized_gates[i]].operation.params[0].name in self.remove_gates:
+                indices_to_be_removed_parameterized_gates.append(indices_parameterized_gates[i])
+
+        assert len(set(indices_to_be_removed_parameterized_gates)) == len({elem for elem in self.remove_gates if elem})
+        return indices_to_be_removed_parameterized_gates
```

Comment:
Took me a while to find out what this is supposed to do. If I got it correctly, this goes through the compiled circuit and marks gates that contain a parameter that is marked for removal in the original circuit.

This somewhat assumes that merely removing the compiled parametrised gates is enough to make the rest of the compiled gates no-ops. While this certainly holds for the RZZ gate, this is kind of a risky assumption for general gates I'd guess. Something we should keep in mind

Reply:
Yeah. But I guess for the prototypical implementation here, it's fine to do it like this.